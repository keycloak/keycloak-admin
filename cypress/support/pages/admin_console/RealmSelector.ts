const expect = chai.expect;

export default class RealmSelector {
  private realmSelector = "realmSelector";
  private realmContextSelector = ".keycloak__realm_selector__context_selector";

  shouldContainAll(realmsList: string[]) {
    cy.getId(this.realmSelector)
      .scrollIntoView()
      .get("ul")
      .should((realms) => {
        for (let index = 0; index < realmsList.length; index++) {
          const realmName = realmsList[index];
          expect(realms).to.contain(realmName);
        }
      });

    return this;
  }

  openRealmContextSelector() {
    cy.getId(this.realmSelector).scrollIntoView();
    cy.get(this.realmContextSelector).click();

    return this;
  }
}
