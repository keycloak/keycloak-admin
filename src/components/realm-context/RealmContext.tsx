import React, { useState, useContext } from "react";
import { WhoAmIContext } from "../../whoami/WhoAmI";

export const RealmContext = React.createContext({
  realm: "",
  setRealm: (realm: string) => {},
});

type RealmContextProviderProps = { children: React.ReactNode };

export const RealmContextProvider = ({
  children,
}: RealmContextProviderProps) => {
  const homeRealm = useContext(WhoAmIContext).getHomeRealm();
  const [realm, setRealm] = useState(homeRealm);

  return (
    <RealmContext.Provider value={{ realm, setRealm }}>
      {children}
    </RealmContext.Provider>
  );
};
