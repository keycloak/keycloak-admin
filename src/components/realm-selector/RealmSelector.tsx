import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  Dropdown,
  DropdownToggle,
  DropdownItem,
  Button,
} from '@patternfly/react-core';

import style from './realm-selector.module.css';

type RealmSelectorProps = {
  realm: string;
  realmList: string[];
};

export const RealmSelector = ({ realm, realmList }: RealmSelectorProps) => {
  const [open, setOpen] = useState(false);
  const history = useHistory();
  const dropdownItems = realmList.map((r) => (
    <DropdownItem component="a" href="/" key={r}>
      {r}
    </DropdownItem>
  ));
  return (
    <Dropdown
      id="realm-select"
      className={style.dropdown}
      isOpen={open}
      toggle={
        <DropdownToggle
          id="realm-select-toggle"
          onToggle={() => setOpen(!open)}
          className={style.toggle}
        >
          {realm}
        </DropdownToggle>
      }
      dropdownItems={[
        ...dropdownItems,
        <DropdownItem component="div" key="add">
          <Button onClick={() => history.push('/add-realm')}>Add Realm</Button>
        </DropdownItem>,
      ]}
    />
  );
};
