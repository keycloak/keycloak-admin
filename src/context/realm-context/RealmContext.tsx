import React, { useContext, useEffect, useState } from "react";
import _ from "lodash";

import RealmRepresentation from "keycloak-admin/lib/defs/realmRepresentation";
import { RecentUsed } from "../../components/realm-selector/recent-used";
import { useErrorHandler } from "react-error-boundary";
import { asyncStateFetch, useAdminClient } from "../auth/AdminClient";
import { WhoAmIContext } from "../whoami/WhoAmI";

type RealmContextType = {
  realm: string;
  setRealm: (realm: string) => void;
  realms: RealmRepresentation[];
  refresh: () => Promise<void>;
};

export const RealmContext = React.createContext<RealmContextType>({
  realm: "",
  setRealm: () => {},
  realms: [],
  refresh: () => Promise.resolve(),
});

type RealmContextProviderProps = { children: React.ReactNode };

export const RealmContextProvider = ({
  children,
}: RealmContextProviderProps) => {
  const { whoAmI } = useContext(WhoAmIContext);
  const [realm, setRealm] = useState(whoAmI.getHomeRealm());
  const [realms, setRealms] = useState<RealmRepresentation[]>([]);
  const adminClient = useAdminClient();
  const errorHandler = useErrorHandler();
  const recentUsed = new RecentUsed();

  const updateRealmsList = (realms: RealmRepresentation[]) => {
    setRealms(_.sortBy(realms, "realm"));
    recentUsed.clean(realms.map((r) => r.realm!));
  };

  useEffect(
    () =>
      asyncStateFetch(
        () => adminClient.realms.find(),
        (realms) => updateRealmsList(realms),
        errorHandler
      ),
    []
  );

  const set = (realm: string) => {
    if (
      realms.length === 0 ||
      realms.findIndex((r) => r.realm == realm) !== -1
    ) {
      recentUsed.setRecentUsed(realm);
      setRealm(realm);
      adminClient.setConfig({
        realmName: realm,
      });
    }
  };
  return (
    <RealmContext.Provider
      value={{
        realm,
        setRealm: set,
        realms,
        refresh: async () => {
          const list = await adminClient.realms.find();
          updateRealmsList(list);
        },
      }}
    >
      {children}
    </RealmContext.Provider>
  );
};

export const useRealm = () => useContext(RealmContext);
