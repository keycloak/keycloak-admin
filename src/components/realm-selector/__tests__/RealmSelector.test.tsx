import React from "react";
import { mount } from "enzyme";
import { act } from "@testing-library/react";

import { RealmSelector } from "../RealmSelector";
import { RealmContextProvider } from "../../../context/realm-context/RealmContext";
import { MemoryRouter } from "react-router-dom";

it("renders realm selector", async () => {
  const wrapper = mount(
    <MemoryRouter>
      <RealmContextProvider>
        <div id="realm">
          <RealmSelector realmList={[{ id: "321", realm: "another" }]} />
        </div>
      </RealmContextProvider>
    </MemoryRouter>
  );

  const expandButton = wrapper.find("button");
  act(() => {
    expandButton!.simulate("click");
  });

  expect(wrapper.find("#realm")).toMatchSnapshot();
});
