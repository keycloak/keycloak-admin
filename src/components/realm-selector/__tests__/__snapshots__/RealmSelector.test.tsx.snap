// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders realm selector 1`] = `
<RealmContextProvider>
  <RealmSelector
    realmList={
      Array [
        Object {
          "id": "321",
          "realm": "another",
        },
      ]
    }
  >
    <Dropdown
      className="keycloak__realm_selector__dropdown"
      dropdownItems={
        Array [
          <DropdownItem
            onClick={[Function]}
          >
            <RealmText
              value="another"
            />
          </DropdownItem>,
          <React.Fragment />,
        ]
      }
      id="realm-select"
      isOpen={false}
      toggle={
        <DropdownToggle
          className="keycloak__realm_selector_dropdown__toggle"
          id="realm-select-toggle"
          onToggle={[Function]}
        >
          Master
        </DropdownToggle>
      }
    >
      <DropdownWithContext
        autoFocus={true}
        className="keycloak__realm_selector__dropdown"
        direction="down"
        dropdownItems={
          Array [
            <DropdownItem
              onClick={[Function]}
            >
              <RealmText
                value="another"
              />
            </DropdownItem>,
            <React.Fragment />,
          ]
        }
        id="realm-select"
        isGrouped={false}
        isOpen={false}
        isPlain={false}
        menuAppendTo="inline"
        onSelect={[Function]}
        position="left"
        toggle={
          <DropdownToggle
            className="keycloak__realm_selector_dropdown__toggle"
            id="realm-select-toggle"
            onToggle={[Function]}
          >
            Master
          </DropdownToggle>
        }
      >
        <div
          className="pf-c-dropdown keycloak__realm_selector__dropdown"
          data-ouia-component-id="OUIA-Generated-Dropdown-1"
          data-ouia-component-type="PF4/Dropdown"
          data-ouia-safe={true}
          id="realm-select"
        >
          <DropdownToggle
            aria-haspopup={true}
            className="keycloak__realm_selector_dropdown__toggle"
            getMenuRef={[Function]}
            id="realm-select-toggle"
            isOpen={false}
            isPlain={false}
            key=".0"
            onEnter={[Function]}
            onToggle={[Function]}
            parentRef={
              Object {
                "current": <div
                  class="pf-c-dropdown pf-m-expanded keycloak__realm_selector__dropdown"
                  data-ouia-component-id="OUIA-Generated-Dropdown-1"
                  data-ouia-component-type="PF4/Dropdown"
                  data-ouia-safe="true"
                  id="realm-select"
                >
                  <button
                    aria-expanded="true"
                    aria-haspopup="true"
                    class="pf-c-dropdown__toggle keycloak__realm_selector_dropdown__toggle"
                    id="realm-select-toggle"
                    type="button"
                  >
                    <span
                      class="pf-c-dropdown__toggle-text"
                    >
                      Master
                    </span>
                    <span
                      class="pf-c-dropdown__toggle-icon"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </button>
                  <ul
                    aria-labelledby="realm-select-toggle"
                    class="pf-c-dropdown__menu"
                    role="menu"
                  >
                    <li
                      role="menuitem"
                    >
                      <a
                        aria-disabled="false"
                        class="pf-c-dropdown__menu-item"
                        tabindex="-1"
                      >
                        <div
                          class="pf-l-split keycloak__realm_selector__list-item-split"
                        >
                          <div
                            class="pf-l-split__item pf-m-fill"
                          >
                            Another
                          </div>
                          <div
                            class="pf-l-split__item"
                          />
                        </div>
                      </a>
                    </li>
                  </ul>
                </div>,
              }
            }
          >
            <Toggle
              aria-haspopup={true}
              bubbleEvent={false}
              className="keycloak__realm_selector_dropdown__toggle"
              getMenuRef={[Function]}
              id="realm-select-toggle"
              isActive={false}
              isDisabled={false}
              isOpen={false}
              isPlain={false}
              isPrimary={false}
              isSplitButton={false}
              onEnter={[Function]}
              onToggle={[Function]}
              parentRef={
                Object {
                  "current": <div
                    class="pf-c-dropdown pf-m-expanded keycloak__realm_selector__dropdown"
                    data-ouia-component-id="OUIA-Generated-Dropdown-1"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                    id="realm-select"
                  >
                    <button
                      aria-expanded="true"
                      aria-haspopup="true"
                      class="pf-c-dropdown__toggle keycloak__realm_selector_dropdown__toggle"
                      id="realm-select-toggle"
                      type="button"
                    >
                      <span
                        class="pf-c-dropdown__toggle-text"
                      >
                        Master
                      </span>
                      <span
                        class="pf-c-dropdown__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </button>
                    <ul
                      aria-labelledby="realm-select-toggle"
                      class="pf-c-dropdown__menu"
                      role="menu"
                    >
                      <li
                        role="menuitem"
                      >
                        <a
                          aria-disabled="false"
                          class="pf-c-dropdown__menu-item"
                          tabindex="-1"
                        >
                          <div
                            class="pf-l-split keycloak__realm_selector__list-item-split"
                          >
                            <div
                              class="pf-l-split__item pf-m-fill"
                            >
                              Another
                            </div>
                            <div
                              class="pf-l-split__item"
                            />
                          </div>
                        </a>
                      </li>
                    </ul>
                  </div>,
                }
              }
            >
              <button
                aria-expanded={false}
                aria-haspopup={true}
                className="pf-c-dropdown__toggle keycloak__realm_selector_dropdown__toggle"
                disabled={false}
                id="realm-select-toggle"
                onClick={[Function]}
                onKeyDown={[Function]}
                type="button"
              >
                <span
                  className="pf-c-dropdown__toggle-text"
                >
                  Master
                </span>
                <span
                  className="pf-c-dropdown__toggle-icon"
                >
                  <CaretDownIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style={
                        Object {
                          "verticalAlign": "-0.125em",
                        }
                      }
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                      />
                    </svg>
                  </CaretDownIcon>
                </span>
              </button>
            </Toggle>
          </DropdownToggle>
        </div>
      </DropdownWithContext>
    </Dropdown>
  </RealmSelector>
</RealmContextProvider>
`;
